src/Microsoft.AspNetCore.Routing/Patterns/RoutePattern.cs

```diff
@@ -0,0 +1,94 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+
+namespace Microsoft.AspNetCore.Routing.Patterns
+{
+    [DebuggerDisplay("{DebuggerToString()}")]
+    public sealed class RoutePattern
```

Comment:
This is a good place to start. There's a set of `RoutePattern*` classes introduced by this PR that make up the new object model.

Reply:
If we're going to make a new thing public, we should make sure it's using all the right types (Span etc).

Reply:
Route patterns (templates) are created once on startup and then live for the lifetime of the app. I don't see Span providing any significant benefit.

RoutePattern should expose whatever is the fastest for matches to be made with. I'm guessing that is `string` but @rynowak has been working on the optimizing matching and would know best.

Reply:
The fact that it's a public API makes me think about usages outside of ours (once at startup).

Reply:
this code plays no role in URL matching